documentation/projects/proposals/trust_and_safety/content_report_moderation/20240122-implementation_plan_moderation_metrics.md

```diff
@@ -0,0 +1,485 @@
+[custom views](https://docs.djangoproject.com/en/5.0/ref/contrib/admin/#adding-views-to-admin-sites)
+and
+[overriding/extending existing templates](https://docs.djangoproject.com/en/5.0/ref/contrib/admin/#overriding-admin-templates).
+These approaches can be coupled together for a good experience in displaying
+these metrics to the moderators.
+
+Values related to image reports can be provided by editing the template that
+Django admin uses to render the "ImageReport" list view. Similar values for
+audio can be shown on the "AudioReport" view.
+
+#### Lists
+
+For the list of the most reported creator and most reported source, we can
+create pseudo-models with `managed = False`, register the model in the Django
+admin site and finally create custom filters to operate on the in-memory list.
+
+```python
+class Creator(models.Model):
+    # fields to uniquely identify a creator
+    name = models.CharField(max_length=100)
+    provider = models.CharField(max_length=100)
+
+    class Meta:
+        managed = False
+
+class CreatorAdmin(admin.ModelAdmin):
+    list_display = ('name', 'provider', 'report_count')
+    list_filter = ('provider',)
+
+    def report_count(self, obj):
+        return Report.objects.filter(media_provider=obj.provider).filter(media__creator=obj.name).count()
+
+    def get_queryset(self, request):
+        qs = super().get_queryset(request)
+        # Convert creators from media models into `Creator` models
+        return Media.objects.values('creator', 'provider').map(lambda x: Creator(name=x['creator'], provider=x['provider'])
+
+class CustomFilter(admin.SimpleListFilter):
+    title = 'Filter Title'
+    parameter_name = 'parameter'
+
+    def lookups(self, request, model_admin):
+        # Return a list of tuples (value, verbose_value) for the filter.
+        # For example, in the provider filter, these can be a list of providers for that media type.
+        return (
+            ('option_one', 'Option 1'),
+            ('option_two', 'Option 2'),
+        )
+
+    def queryset(self, request, queryset):
+        # Manipulate the pseudo-queryset from `get_queryset` using `filter` or list comprehensions
+        # Use self.value() which will be the 0-th element of the tuple returned by `lookups`
+        return queryset
+```
+
+For the lists of most reported media item, we register the media model again
+using a proxy as Django disallows one model to be registered more than once.
+
+```python
+class ImageProxy(Image):
+    class Meta:
+        proxy = True
+
+class AudioProxy(Audio):
+    class Meta:
+        proxy = True
+```
+
+## Dependencies
+
+This project should introduce no new dependencies as this builds on top of our
+existing technology stack.
+
+### Infrastructure
+
+We already use Logs Insights, CloudWatch and Django Admin in our stack and, more
+specifically, in the moderation pipeline.
+
+### Other projects or work
+
+This project depends on our implementation of the moderation workflow. By having
```

Comment:
Do you mind linking the specific projects that are dependent here?

Reply:
Linked in 3a5fda2.